pallets/foreign-investments/src/types.rs

```diff
@@ -0,0 +1,330 @@
+	ActiveSwapIntoPoolCurrency { swap: Swap<Balance, Currency> },
+	/// The unprocessed investment was fully decreased and is currently swapped
+	/// back into the corresponding foreign currency.
+	ActiveSwapIntoForeignCurrency { swap: Swap<Balance, Currency> },
+	/// The investment is not fully swapped into pool currency and thus split
+	/// into two parts:
+	/// * One part is still being swapped.
+	/// * The remainder is already waiting to be processed as investment.
+	ActiveSwapIntoPoolCurrencyAndInvestmentOngoing {
+		swap: Swap<Balance, Currency>,
+		invest_amount: Balance,
+	},
+	/// The investment is split into two parts:
+	/// * One part is waiting to be processed as investment.
+	/// * The remainder is swapped back into the foreign currency as a result of
+	///   decreasing the invested amount before being processed.
+	ActiveSwapIntoForeignCurrencyAndInvestmentOngoing {
+		swap: Swap<Balance, Currency>,
+		invest_amount: Balance,
+	},
+	/// The investment is split into two parts:
+	/// * The one part is swapping into pool currency.
+	/// * The remainder was swapped back into the foreign currency as a result
+	///   of decreasing the invested amount before being processed.
+	///
+	/// NOTE: This state can be transitioned into `ActiveSwapIntoPoolCurrency`
+	/// by applying the corresponding trigger to handle the foreign return
+	/// amount.
+	ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDone {
+		swap: Swap<Balance, Currency>,
+		done_amount: Balance,
+	},
+	/// The investment is swapped back into the foreign currency and was already
+	/// partially fulfilled.
+	ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+		swap: Swap<Balance, Currency>,
+		done_amount: Balance,
+	},
+	/// The investment is split into three parts:
+	/// * One part is currently swapping into the pool currency.
+	/// * The second is already waiting to be processed as investment.
+	/// * The remainder was swapped back into the foreign currency as a result
+	///   of decreasing the invested amount before being processed.
+	///
+	/// NOTE: This state can be transitioned into
+	/// `ActiveSwapIntoPoolCurrencyAndInvestmentOngoing` by applying the
+	/// corresponding trigger to handle the foreign return amount.
+	ActiveSwapIntoPoolCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
+		swap: Swap<Balance, Currency>,
+		done_amount: Balance,
+		invest_amount: Balance,
+	},
+	/// The investment is split into three parts:
+	/// * One is waiting to be processed as investment.
+	/// * The second is swapped back into the foreign currency as a result of
+	///   decreasing the invested amount before being processed.
+	/// * The remainder was already swapped back into the foreign currency.
+	///
+	/// NOTE: This state should not be transitioned by applying the trigger for
+	/// the done part but wait until the active swap is fulfilled.
+	ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDoneAndInvestmentOngoing {
```

Comment:
Super nit comment (totally skippable). I think maybe this name could be shortened, to something like: `SwappingAndSwapedAndInvested`.

Along with this, I think `impl` modules, by convention, are used for traits implemented by the pallet. So maybe these `invest` and `redeem` implementations, which are related to independent types, could then be placed under `types`: `types/invest.rs` and `types/redeem.rs`.

(everything super nit and not mandatory at all)

Reply:
Thanks for pointing this out! I don't want to rename the states at this point but will add this to the list of potential post-PR improvements. Same for the reordering of the impls. Both would create noise and make the review process even more difficult.